Use mem::MaybeUninit instead of mem::uninitialized() since Rust 1.36

rhysd opened the pull request:
I noticed that CI fails since this crate is using deprecated `std::mem::uninitialized()`.

https://ci.appveyor.com/project/Stebalien/term/builds/26853323/job/880r4cnu5lhfaccv

This PR uses [`MaybeUninit`](https://doc.rust-lang.org/beta/std/mem/union.MaybeUninit.html) which replaces `std::mem::uninitialized()` using a compiler version switch provided by [`rustversion` crate](https://docs.rs/rustversion/0.1.4/rustversion/).

Stebalien commented on src/win.rs:
```diff
@@ -161,6 +161,15 @@ fn test_conout() {
+}
+#[rustversion::since(1.36)]
+unsafe fn uninitialized<T>() -> T {
+    ::std::mem::MaybeUninit::uninit().assume_init()
```
This is undefined behavior. `assume_init` can't be called until we actually initialize it with `GetConsoleScreenBufferInfo`.

rhysd replied:
I thought an uninitialized value was ok since the value will be initialized by the function soon. But as you pointed, returning uninitialized buffer may cause undefined behavior and we should eliminate it.

<!--
How about initializing the buffer with zero?

```rust
#[rustversion::since(1.36)]
unsafe fn uninitialized<T>() -> T {
    ::std::mem::MaybeUninit::zeroed().assume_init()
}
```

Since every bit is initialized to zero, the behavior won't be undefined. Though I read a documentation of [the buffer info struct](https://docs.microsoft.com/ja-jp/windows/console/console-screen-buffer-info-str), I'm not sure filling with zero is sufficient.-->

rhysd commented:
I have addressed the review comment at 8edc0474dd4d27b0af055feefb583794dad4b3f2. Instead of `uninitialized()`, it wraps `GetConsoleScreenBufferInfo()` as `get_console_screen_buffer_info()`. After getting uninitialized memory, it immediately initializes the buffer with `GetConsoleScreenBufferInfo()`. So the returned buffer is guaranteed to be initialized and there is no undefined behavior.

(I apologize many force pushes. I don't have Windows machine so I needed to check Appveyor result instead)

Stebalien commented on src/win.rs:
```diff
@@ -161,6 +163,25 @@ fn test_conout() {
+}
+#[rustversion::since(1.36)]
+unsafe fn get_console_screen_buffer_info(handle: HANDLE) -> io::Result<CONSOLE_SCREEN_BUFFER_INFO> {
+    let mut buffer_info = ::std::mem::MaybeUninit::uninit().assume_init();
```
This needs to call `assume_init` _after_ filling the buffer. That is, it should call `GetConsoleScreenBufferInfo(handle, buffer_info.as_mut_ptr()` then `buffer_info.assume_init()` (if there was no error).

We can't take a reference to uninitialized memory (which is why this type was introduced).

rhysd commented:
Ahh, ok, I misunderstood the semantics of `MaybeUninit`. Thank you for the point.

rhysd commented:
I fixed my patch again at 4fc39a93 and add a new small test for `get_console_screen_buffer_info()` at 65801a0.

rhysd commented:
Thank you for correcting me and helping to understand correctly.
